File: lld/problems/rate-limiter/go/rate_limiter.go
// Package ratelimiter implements a single-process, per-client rate limiter
// with two interchangeable algorithms — token bucket and sliding window —
// behind a common RateLimiter interface. See ../README.md for the design
// writeup this package implements.
package ratelimiter

import (
	"sync"
	"time"
)

// RateLimiter is the common strategy interface every algorithm implements,
// so calling code never branches on which algorithm is active.
type RateLimiter interface {
	// Allow decides whether a request from clientID is permitted right now.
	Allow(clientID string) bool
}

// clock returns the current time; it is a field (not a package var) on each
// limiter so tests can inject a fake clock instead of sleeping real time.
type clock func() time.Time

// bucket holds one client's token-bucket state.
type bucket struct {
	tokens     float64
	lastRefill time.Time
}

// TokenBucketLimiter allows bursts up to capacity tokens, refilling at
// refillRate tokens/second. Refill is computed lazily on each Allow call
// rather than via a background goroutine per client.
type TokenBucketLimiter struct {
	capacity   float64
	refillRate float64 // tokens per second
	now        clock

	mu      sync.Mutex
	buckets map[string]*bucket
}

// NewTokenBucketLimiter builds a token bucket limiter with the given
// capacity (max burst size) and refillRate (tokens added per second).
func NewTokenBucketLimiter(capacity, refillRate float64) *TokenBucketLimiter {
	return &TokenBucketLimiter{
		capacity:   capacity,
		refillRate: refillRate,
		now:        time.Now,
		buckets:    make(map[string]*bucket),
	}
}

// refill tops b up based on elapsed time since its last refill, capping at
// capacity. Caller must hold l.mu.
func (l *TokenBucketLimiter) refill(b *bucket, now time.Time) {
	elapsed := now.Sub(b.lastRefill).Seconds()
	if elapsed <= 0 {
		return
	}
	b.tokens += elapsed * l.refillRate
	if b.tokens > l.capacity {
		b.tokens = l.capacity
	}
	b.lastRefill = now
}

// Allow consumes one token for clientID if available.
func (l *TokenBucketLimiter) Allow(clientID string) bool {
	l.mu.Lock()
	defer l.mu.Unlock()

	now := l.now()
	b, ok := l.buckets[clientID]
	if !ok {
		b = &bucket{tokens: l.capacity, lastRefill: now}
		l.buckets[clientID] = b
	} else {
		l.refill(b, now)
	}

	if b.tokens < 1 {
		return false
	}
	b.tokens--
	return true
}

// window holds one client's sliding-log state: timestamps of requests still
// inside the trailing window.
type window struct {
	timestamps []time.Time
}

// SlidingWindowLimiter allows at most limit requests in any trailing
// duration of length window (a sliding log), giving a smooth, hard cap with
// no boundary burst allowance.
type SlidingWindowLimiter struct {
	limit  int
	window time.Duration
	now    clock

	mu      sync.Mutex
	windows map[string]*window
}

// NewSlidingWindowLimiter builds a sliding window limiter allowing at most
// limit requests within any trailing duration of the given window size.
func NewSlidingWindowLimiter(limit int, windowSize time.Duration) *SlidingWindowLimiter {
	return &SlidingWindowLimiter{
		limit:   limit,
		window:  windowSize,
		now:     time.Now,
		windows: make(map[string]*window),
	}
}

// Allow evicts timestamps older than the window and checks the remaining
// count against limit, recording the new request if it fits.
func (l *SlidingWindowLimiter) Allow(clientID string) bool {
	l.mu.Lock()
	defer l.mu.Unlock()

	now := l.now()
	w, ok := l.windows[clientID]
	if !ok {
		w = &window{}
		l.windows[clientID] = w
	}

	cutoff := now.Add(-l.window)
	kept := w.timestamps[:0]
	for _, ts := range w.timestamps {
		if ts.After(cutoff) {
			kept = append(kept, ts)
		}
	}
	w.timestamps = kept

	if len(w.timestamps) >= l.limit {
		return false
	}
	w.timestamps = append(w.timestamps, now)
	return true
}

// RateLimiterRegistry maps a client class (e.g. free/paid/enterprise) to the
// RateLimiter instance configured for that tier, so different tiers can even
// run different algorithms.
type RateLimiterRegistry struct {
	mu             sync.RWMutex
	perClientClass map[string]RateLimiter
}

// NewRateLimiterRegistry builds an empty registry.
func NewRateLimiterRegistry() *RateLimiterRegistry {
	return &RateLimiterRegistry{perClientClass: make(map[string]RateLimiter)}
}

// Register associates clientClass with limiter, overwriting any prior
// registration for that class.
func (r *RateLimiterRegistry) Register(clientClass string, limiter RateLimiter) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.perClientClass[clientClass] = limiter
}

// GetLimiter returns the RateLimiter configured for clientClass, and false
// if no limiter has been registered for it.
func (r *RateLimiterRegistry) GetLimiter(clientClass string) (RateLimiter, bool) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	limiter, ok := r.perClientClass[clientClass]
	return limiter, ok
}

File: lld/problems/rate-limiter/python/rate_limiter.py
"""Rate Limiter LLD — Python reference implementation.

Two interchangeable per-client algorithms (token bucket, sliding window)
behind a common RateLimiter interface, plus a RateLimiterRegistry that maps
a client class (free/paid/enterprise) to its configured limiter. See
../README.md for the design writeup.
"""
from __future__ import annotations

import threading
import time
from abc import ABC, abstractmethod
from collections import deque
from typing import Callable, Deque, Dict, Optional


class RateLimiter(ABC):
    """Common strategy interface every algorithm implements, so calling
    code never branches on which algorithm is active."""

    @abstractmethod
    def allow(self, client_id: str) -> bool:
        """Decides whether a request from client_id is permitted right now."""
        raise NotImplementedError


class _Bucket:
    __slots__ = ("tokens", "last_refill")

    def __init__(self, tokens: float, last_refill: float):
        self.tokens = tokens
        self.last_refill = last_refill


class TokenBucketLimiter(RateLimiter):
    """Allows bursts up to capacity tokens, refilling at refill_rate
    tokens/second. Refill is computed lazily on each allow() call rather
    than via a background thread per client."""

    def __init__(self, capacity: float, refill_rate: float, clock: Callable[[], float] = time.monotonic):
        self._capacity = capacity
        self._refill_rate = refill_rate  # tokens per second
        self._clock = clock
        self._buckets: Dict[str, _Bucket] = {}
        self._lock = threading.Lock()

    def _refill(self, bucket: _Bucket, now: float) -> None:
        elapsed = now - bucket.last_refill
        if elapsed <= 0:
            return
        bucket.tokens = min(self._capacity, bucket.tokens + elapsed * self._refill_rate)
        bucket.last_refill = now

    def allow(self, client_id: str) -> bool:
        with self._lock:
            now = self._clock()
            bucket = self._buckets.get(client_id)
            if bucket is None:
                bucket = _Bucket(self._capacity, now)
                self._buckets[client_id] = bucket
            else:
                self._refill(bucket, now)

            if bucket.tokens < 1:
                return False
            bucket.tokens -= 1
            return True


class _Window:
    __slots__ = ("timestamps",)

    def __init__(self) -> None:
        self.timestamps: Deque[float] = deque()


class SlidingWindowLimiter(RateLimiter):
    """Allows at most `limit` requests in any trailing `window_seconds`
    duration (a sliding log), giving a smooth, hard cap with no boundary
    burst allowance."""

    def __init__(self, limit: int, window_seconds: float, clock: Callable[[], float] = time.monotonic):
        self._limit = limit
        self._window = window_seconds
        self._clock = clock
        self._windows: Dict[str, _Window] = {}
        self._lock = threading.Lock()

    def allow(self, client_id: str) -> bool:
        with self._lock:
            now = self._clock()
            window = self._windows.setdefault(client_id, _Window())

            cutoff = now - self._window
            while window.timestamps and window.timestamps[0] <= cutoff:
                window.timestamps.popleft()

            if len(window.timestamps) >= self._limit:
                return False
            window.timestamps.append(now)
            return True


class RateLimiterRegistry:
    """Maps a client class (e.g. free/paid/enterprise) to the RateLimiter
    instance configured for that tier, so different tiers can even run
    different algorithms."""

    def __init__(self) -> None:
        self._per_client_class: Dict[str, RateLimiter] = {}
        self._lock = threading.Lock()

    def register(self, client_class: str, limiter: RateLimiter) -> None:
        with self._lock:
            self._per_client_class[client_class] = limiter

    def get_limiter(self, client_class: str) -> Optional[RateLimiter]:
        with self._lock:
            return self._per_client_class.get(client_class)


def _demo() -> None:
    token_bucket = TokenBucketLimiter(capacity=2, refill_rate=1)
    print(f"token bucket allow(a) = {token_bucket.allow('a')}")
    print(f"token bucket allow(a) = {token_bucket.allow('a')}")
    print(f"token bucket allow(a) [should block] = {token_bucket.allow('a')}")

    sliding_window = SlidingWindowLimiter(limit=2, window_seconds=1)
    print(f"sliding window allow(b) = {sliding_window.allow('b')}")
    print(f"sliding window allow(b) = {sliding_window.allow('b')}")
    print(f"sliding window allow(b) [should block] = {sliding_window.allow('b')}")


if __name__ == "__main__":
    _demo()
